From 31bf9dcfc618a23ecb72227dfeed5084be05dd32 Mon Sep 17 00:00:00 2001
From: Ming Lu <ming.lu@citrix.com>
Date: Thu, 9 Jan 2020 11:25:04 +0000
Subject: [PATCH] Use linux/random.h

---
 crypto/rand/rand_unix.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/crypto/rand/rand_unix.c b/crypto/rand/rand_unix.c
index 11716f7..44e9cde 100644
--- a/crypto/rand/rand_unix.c
+++ b/crypto/rand/rand_unix.c
@@ -21,7 +21,7 @@
 #include "internal/dso.h"
 #if defined(__linux)
 # include <sys/syscall.h>
-# include <sys/random.h>
+# include <linux/random.h>
 #endif
 #if defined(__FreeBSD__)
 # include <sys/types.h>
-- 
1.8.3.1

